### PR TITLE
Support Codable Messages

### LINKED ISSUE #62: Codable messages
Would be great to be able to specify a generic `Message` type on `client.connect`, that would auto encode/decode if `Codable` conformance is present. This type could default to `String`.

`onText` could also be renamed `onMessage`.

### PR DESCRIPTION
**Pull request by dnKaratzas**
(#103,  fixes #62).

Many Thanks to @MihaelIsaev  for the [inspiration](https://github.com/MihaelIsaev/AwesomeWS)!

Those changes adds support to send `Codable` messages and receive events in special JSON format:

`{ "event": "<event name>" }`
or
`{ "event": "<event name>", "data": <anything> }`

The event `data` will be decoded to the provided `Type`.

Here is a demo:
```swift
struct WebSocketController: RouteCollection {
	func boot(routes: RoutesBuilder) throws {
		routes.webSocket("chat") { req, ws in
			ws.onEvent("hello", helloHandler)

			ws.onEvent("bye", Bye.self) { ws, data in
				print("Bye \(data.firstName) \(data.lastName)")
				ws.send(Response(status: "Bye bye!"))
			}

			ws.onText { ws, text in
				print(text)
			}

			ws.onEvent("stop") { ws in
				// stop event without data
			}

			ws.onEvent("start", startHandler)
		}
	}
}

// MARK: - WebSocket Event Handlers
extension WebSocketController {
	func helloHandler(webSocket: WebSocket, data: Hello) {
		print("Hello \(data.firstName) \(data.lastName)")
		webSocket.send(Response(status: "Hey There \(data.firstName) \(data.lastName)"), type: .binary)
	}

	func startHandler(webSocket: WebSocket) {
		// start event
	}
}

// MARK: - Codables
struct Hello: Codable {
	let firstName, lastName: String
}

struct Bye: Codable {
	let firstName, lastName: String
}

struct Response: Codable {
	let status: String
}
```

The code will become much cleaner as you can give on events a handler!

### CODE REVIEW

File: Sources/WebSocketKit/WebSocket.swift [hunk 1]
```diff
@@ -63,6 +75,32 @@ public final class WebSocket {
...
+
+    public func onEvent<T>(_ identifier: String, _ type: T.Type, _ handler: @escaping (WebSocket, T) -> Void) where T: Codable {
+        onEvent(identifier, handler)
+    }
+
+    public func onEvent<T>(_ identifier: String, _ handler: @escaping (WebSocket, T) -> Void) where T: Codable {
+        events[identifier] = { [weak self] ws, data in
+            guard let self = self else { return }
+            do {
+                let res = try JSONDecoder().decode(WebSocketEvent<T>.self, from: data)
```
Comment (0xTim) [inline Sources/WebSocketKit/WebSocket.swift:92]: We should use the decoder from above
```suggestion
                let res = try self.decoder.decode(WebSocketEvent<T>.self, from: data)
```

File: Sources/WebSocketKit/WebSocket.swift [hunk 2]
```diff
@@ -24,7 +30,11 @@ public final class WebSocket {
         self.channel.closeFuture
     }
 
+    public var jsonDecoder = JSONDecoder()
+    public var jsonEncoder = JSONEncoder()
+
```
Comment (0xTim) [inline Sources/WebSocketKit/WebSocket.swift:35]: Lets keep these private for now. If we want them to be configurable we should instead allow them to be passed in to the methods where they are used and default to new ones

File: Sources/WebSocketKit/Exports.swift [hunk 3]
```diff
@@ -8,3 +8,5 @@
 @_exported import struct NIOHTTP1.HTTPHeaders
 
 @_exported import struct Foundation.URL
+
+@_exported import Logging
```
Comment (0xTim) [inline Sources/WebSocketKit/Exports.swift:12]: We shouldn't add this as it makes the whole Logging module part of our public API (I know we've done this for other stuff but that was a mistake 😆 )

```suggestion
@_exported import struct Foundation.URL
```

File: Sources/WebSocketKit/WebSocket.swift [hunk 4]
```diff
@@ -63,6 +75,32 @@ public final class WebSocket {
...
+            handler(ws)
+        }
+    }
+
+    public func onEvent<T>(_ identifier: String, _ type: T.Type, _ handler: @escaping (WebSocket, T) -> Void) where T: Codable {
+        onEvent(identifier, handler)
+    }
+
+    public func onEvent<T>(_ identifier: String, _ handler: @escaping (WebSocket, T) -> Void) where T: Codable {
+        events[identifier] = { [weak self] ws, data in
```
Comment (0xTim) [inline Sources/WebSocketKit/WebSocket.swift:89]: Why are we using a weak self here?

File: Sources/WebSocketKit/WebSocket.swift [hunk 5]
```diff
@@ -63,6 +75,32 @@ public final class WebSocket {
...
+            guard let self = self else { return }
+            do {
+                let res = try JSONDecoder().decode(WebSocketEvent<T>.self, from: data)
+                if let data = res.data {
+                    handler(ws, data)
+                } else {
+                    self.logger.warning("Unable to unwrap data for event `\(identifier)`, because it is unexpectedly nil. Please use another `bind` method which support optional payload to avoid this message.")
```
Comment (0xTim) [inline Sources/WebSocketKit/WebSocket.swift:96]: This should be trace most likely. Warnings should rarely be thrown by libraries. See the [Logging Guide](https://github.com/swift-server/guides/blob/main/docs/libs/log-levels.md) for more details

File: Sources/WebSocketKit/WebSocket.swift [hunk 6]
```diff
@@ -63,6 +75,32 @@ public final class WebSocket {
...
+                }
+            } catch {
+                self.logger.error("Unable to decode incoming event `\(identifier)`: \(error)")
```
Comment (0xTim) [inline Sources/WebSocketKit/WebSocket.swift:99]: This should probably be debug.  See above - errors are the kind of thing you get people out of bed for!

File: Sources/WebSocketKit/WebSocket.swift [hunk 7]
```diff
@@ -256,6 +310,37 @@ public final class WebSocket {
...
+            var bytes: [UInt8] = byteBuffer.getBytes(at: byteBuffer.readerIndex, length: byteBuffer.readableBytes) ?? []
+            let data = Data(bytes: &bytes, count: byteBuffer.readableBytes)
+
+            let prototype = try jsonDecoder.decode(WebSocketEventPrototype.self, from: data)
+            if let bind = events.first(where: { $0.0 == prototype.event }) {
+                bind.value(socket, data)
+                return true
+            }
+        } catch {
+            logger.trace("Unable to decode incoming event cause it doesn't conform to `WebSocketEventPrototype` model: \(error)")
```
Comment (0xTim) [inline Sources/WebSocketKit/WebSocket.swift:339]: Nit spelling
```suggestion
            logger.trace("Unable to decode incoming event because it doesn't conform to `WebSocketEventPrototype` model: \(error)")
```

File: Sources/WebSocketKit/WebSocket.swift [hunk 8]
```diff
@@ -311,3 +396,16 @@ private struct WebSocketFrameSequence {
...
+    public let event: String
+    public let data: T?
+    public init (event: String, data: T? = nil) {
+        self.event = event
+        self.data = data
+    }
+}
+
+private struct WebSocketEventPrototype: Codable {
+    public var event: String
```
Comment (0xTim) [inline Sources/WebSocketKit/WebSocket.swift:410]: These don't need to be public if the type is private

File: Tests/WebSocketKitTests/AsyncWebSocketKitTests.swift [hunk 9]
```diff
@@ -42,10 +43,56 @@ final class AsyncWebSocketKitTests: XCTestCase {
...
+            return
+        }
+
+        let promise = elg.next().makePromise(of: String.self)
+
+        try await WebSocket.connect(to: "ws://localhost:\(port)", on: elg) { ws in
+            do {
+                try await ws.send(Response(event: "hello", data: User(firstName: "Vapor", lastName: "WebSocket")))
+                ws.onText { ws, string in
+                    promise.succeed(string)
```
Comment (0xTim) [inline Tests/WebSocketKitTests/AsyncWebSocketKitTests.swift:64]: We shouldn't be touching promises in async tests

**Review by 0xTim — CHANGES_REQUESTED** [review]
@dnKaratzas I've added the logging APIs needed to be able to inject it in. A few things to discuss but thanks for the PR!

File: Sources/WebSocketKit/WebSocket.swift (hunk 2 above)
Comment (dnKaratzas) [reply Sources/WebSocketKit/WebSocket.swift:35]: Ok changed them to private

File: Sources/WebSocketKit/Exports.swift (hunk 3 above)
Comment (dnKaratzas) [reply Sources/WebSocketKit/Exports.swift:12]: Agree with you. Reverted the change

File: Sources/WebSocketKit/WebSocket.swift (hunk 4 above)
Comment (dnKaratzas) [reply Sources/WebSocketKit/WebSocket.swift:89]: Reverted!

File: Tests/WebSocketKitTests/AsyncWebSocketKitTests.swift (hunk 9 above)
Comment (dnKaratzas) [reply Tests/WebSocketKitTests/AsyncWebSocketKitTests.swift]: I get the template of that test from the already existing test above it.  Btw changed it...

**Review by 0xTim — APPROVED** [review]
LGTM!

**Review by Joannis — CHANGES_REQUESTED** [review]
I love this! But I want to have a good look at this before it gets merged.

First of all, sending and receiving JSON as text is a very common and valid use case. But I'd prefer that this API doesn't imply that 'binary' means 'UTF-8 encoded JSON Codable types'. If I'm using binary, which doesn't happen much anyways, I'm probably not sending JSON Codable types. I'm using a binary format of something.

I would love to not have all unnecessary APIs added. Specifically the (binary) WebSocketSendType. But also the event types, because I don't think there's a 'standard' for websocket events anywhere. And if there is, this should be documented as being said standard.

Also, the implementation assumes that an event's contents are optional. For me, that would _never_ be the case, unless I'd mark the event as `WebSocketevent<MyEvent?>`.

**In Short:**

I love Codable messages, not so fond of adding the rest to something like WebSocketKit.
And while sending Codable data is supported, receiving Codable data only works on the `Event` type. I'd prefer seeing an `onJSONText(MyEvent.self) { ws, event in .. }`

**Comment by dnKaratzas** [comment]
@Joannis  could you provide your suggestions? Cause I really don't get how do you want it. Commit comment etc ....
